src/tamp_improv/approaches/improvisational/graph_training.py

```diff
@@ -511,6 +531,131 @@ def identify_shortcut_candidates(
     return shortcut_candidates
 
 
+def identify_promising_shortcuts_with_rollouts(
+    system,
+    planning_graph,
+    observed_states: dict[int, Any],
+    num_rollouts_per_node: int = 50,
+    max_steps_per_rollout: int = 30,
+    shortcut_success_threshold: int = 1,
+    random_seed: int = 42,
+) -> list[ShortcutCandidate]:
+    """Identify promising shortcuts by performing random rollouts from each
+    node.
+
+    This function runs random rollouts from each source node in the planning graph
+    and tracks which preimages/nodes are reached during exploration. Shortcuts that
+    are reached at least `shortcut_success_threshold` times through random exploration
+    are considered promising candidates for training.
+    """
+    print("\n=== Identifying Promising Shortcuts with Random Rollouts ===")
+
+    np.random.seed(random_seed)
+
+    shortcut_success_counts: defaultdict[tuple[int, int], int] = defaultdict(
+        int
+    )  # (source_node_id, target_node_id) -> count
+    promising_candidates = []
+
+    # Get the base environment for running rollouts
+    raw_env = system.env
+
+    # For each node with an observed state, perform random rollouts
+    for source_node_id, source_state in observed_states.items():
+        source_node = next(
+            (n for n in planning_graph.nodes if n.id == source_node_id), None
+        )
+        assert source_node is not None
+        source_atoms = set(source_node.atoms)
+        print(
+            f"\nPerforming {num_rollouts_per_node} rollouts from node {source_node_id}"
+        )
+
+        # Track preimages reached from this source node
+        reached_preimages: defaultdict[int, int] = defaultdict(
+            int
+        )  # target_node_id -> count
+
+        # Perform random rollouts
+        for rollout_idx in range(num_rollouts_per_node):
+            if rollout_idx > 0 and rollout_idx % 100 == 0:
+                print(f"  Completed {rollout_idx}/{num_rollouts_per_node} rollouts")
+
+            # Reset the environment to source state
+            raw_env.reset_from_state(source_state)
+            curr_atoms = source_atoms.copy()
+
+            # Execute random actions
+            for _ in range(max_steps_per_rollout):
+                action = raw_env.action_space.sample()
```

Comment:
double check that this is deterministic. if not, you may need to do `raw_env.action_space.seed(seed)`.